tensorflow/compiler/xla/service/gpu/gpu_compiler.cc

```diff
@@ -1176,11 +1176,44 @@ GpuCompiler::CompileToTargetBinary(const HloModuleConfig& module_config,
     }
   }
 
+  // Record the name of some constant global variables and their initializers.
+  // We'll change the linkage type of these variables from external to internal
+  // to ensuare constant-folding works properly after calling llvm::SplitModule.
+  llvm::DenseMap<llvm::StringRef, llvm::Constant*> const_initializer_map;
+  for (llvm::GlobalVariable& gv : llvm_module->globals()) {
+    if (gv.hasName() && gv.isConstant() && gv.hasInitializer() &&
+        gv.hasExternalLinkage()) {
+      llvm::Constant* initializer = gv.getInitializer();
+      unsigned int num_elements = 0;
+      if (auto* caz =
+              llvm::dyn_cast<llvm::ConstantAggregateZero>(initializer)) {
+        num_elements = caz->getElementCount().getFixedValue();
+      }
+      if (auto* cds =
+              llvm::dyn_cast<llvm::ConstantDataSequential>(initializer)) {
+        num_elements = cds->getNumElements();
+      }
+      // Here we only select constant global variables with small number of
+      // elements
+      if (num_elements > 0 && num_elements <= 128) {
```

Comment:
Why 128?

Reply:
The original purpose here is to skip those large constants in case copying their initializers to multiple llvm modules may increase the executable size while not benefiting from constant-folding (not sure if this claim is valid). 128 is a magic number I picked for this purpose.

Based on the models I've tested, only scalar constants like `u64[] constant(32), f32[] constant(0), f16[] constant(0)` in HLO were translated into llvm global constants with num_elements = `8, 4, 2` respectively. Constant arrays like `pred[6]{0} constant({0, 0, 1, 1, 1, 1})` in HLO were translated into non-constant llvm global variable.

So should I remove this check assuming all llvm global constants are scalar types in XLA? Any suggestions?

Reply:
@timshen91 has more context here, deferring to him. Logic in https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc;l=572?q=f:ir_emitter_unnested.cc:572 might be relevant.

Reply:
Yanming, have you actually seen constant global variables that has more than 1 element? XLA is supposed to generate constants only for single-element variables. See https://github.com/tensorflow/tensorflow/blob/04fd9533800eacbf7a07c3c5751b4d874acd3287/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc#L572

Reply:
@timshen91 Thanks for the info. No, I only saw HLO constants with element = 1 like `u64[] constant(32), f32[] constant(0), f16[] constant(0)` in HLO got converted into llvm constant global variables with their corresponding llvm IRs:
```
@buffer_for_constant_143 = constant [8 x i8] c" \00\00\00\00\00\00\00", align 128
@buffer_for_constant_1020 = constant [4 x i8] zeroinitializer, align 128
@buffer_for_constant_5279 = constant [2 x i8] zeroinitializer, align 128
```
Note the number of elements in llvm ir are 8, 4, 2.

HLO constants with elements > 1 like `constant_4431 = pred[512]{0} constant({...})` were translated into 
```
@buffer_for_constant_4775 = global [512 x i8] zeroinitializer, align 128
```
Upon searching the codebase, I only saw [embedded_protocol_buffers.cc](https://github.com/tensorflow/tensorflow/blob/98d7b126cddbb3aa9f93ebc8786ae3e1bcd49bcb/tensorflow/compiler/aot/embedded_protocol_buffers.cc#L53-L56) can potentially generate constant global variable with more than one element whereas [ir_emitter_nested.cc](https://github.com/tensorflow/tensorflow/blob/945f871ce0bfef2a6513e68a5c4e05e36c4d7aa3/tensorflow/compiler/xla/service/gpu/ir_emitter_nested.cc#L224-L230) and [ir_emitter_unested.cc](https://github.com/tensorflow/tensorflow/blob/5f813501f79b0713bbfe87e3b911330fd0f1882f/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc#L600-L606) uses this `should_emit_initializer` check you mentioned.

Reply:
Yeah, for this line I think we can select all constants without the "<= 128" filtering.

Reply:
Sounds good. I've removed it in the revision.